The start of documentation 2.0 (#3673)
CliMA/Oceananigans.jl: docs/src/fields.md

To build such `Fields`, we write

```jldoctest fields
using Oceananigans

grid = RectilinearGrid(size = (4, 4, 4),
                       topology = (Periodic, Periodic, Bounded),
                       x = (0, 1),
                       y = (0, 1),
                       z = (0, 1))

c = Field{Center, Center, Center}(grid)

# output
4×4×4 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 4×4×4 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 10×10×10 OffsetArray(::Array{Float64, 3}, -2:7, -2:7, -2:7) with eltype Float64 with indices -2:7×-2:7×-2:7
    └── max=0.0, min=0.0, mean=0.0
```

Fully-centered fields also go by the alias `CenterField`,

```jldoctest fields
c == CenterField(grid)

# output
true
```

`CenterField`s are some of the most common `Field`s -- in fluid dynamics simulations, for example, tracer quantities like heat and salt are located at cell centers.

> Another common `Field` we encounter at those with finite volumes centered over the `x`-interfaces of the primary grid,

Reviewer: ```suggestion
Another common type of `Field` we encounter are those with finite volumes centered over the `x`-interfaces of the primary grid,
```

To build such `Fields`, we write

```jldoctest fields
using Oceananigans

grid = RectilinearGrid(size = (4, 4, 4),
                       topology = (Periodic, Periodic, Bounded),
                       x = (0, 1),
                       y = (0, 1),
                       z = (0, 1))

c = Field{Center, Center, Center}(grid)

# output
4×4×4 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 4×4×4 RectilinearGrid{Float64, Periodic, Periodic, Bounded} on CPU with 3×3×3 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Periodic, north: Periodic, bottom: ZeroFlux, top: ZeroFlux, immersed: ZeroFlux
└── data: 10×10×10 OffsetArray(::Array{Float64, 3}, -2:7, -2:7, -2:7) with eltype Float64 with indices -2:7×-2:7×-2:7
    └── max=0.0, min=0.0, mean=0.0
```

Fully-centered fields also go by the alias `CenterField`,

```jldoctest fields
c == CenterField(grid)

# output
true
```

`CenterField`s are some of the most common `Field`s -- in fluid dynamics simulations, for example, tracer quantities like heat and salt are located at cell centers.

> A second common `Field` to encounter has finite volumes centered over the `x`-interfaces of the primary grid,

Reviewer: ```suggestion
Another common `Field` we encounter at those with finite volumes centered over the `x`-interfaces of the primary grid,
```

```jldoctest fields
u = Field{Face, Center, Center}(grid)

# output
```

which can also be constructed by writing `u = XFaceField(grid)`.

> The name `u` is suggestive, because within the "C-grid" system used by Oceananigans,

Reviewer: ```suggestion
The name `u` is suggestive, because within the Arakawa type C grid (or, 'C grid' for short) used by Oceananigans,
```

Author: might be good to include a reference

the `x`-component of the velocity field is stored at `Face, Center, Center`.

Staggering means that the "position" of `CenterField` differs from `XFaceField`.
For example, the center of the primary cells have `x`-coordinates

```jldoctest fields
xnodes(c)

# output
4-element view(OffsetArray(::StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}, -2:7), 1:4) with eltype Float64:
 0.125
 0.375
 0.625
 0.875
```

whereas the center of the `XFaceField` cells are located at

```jldoctest fields
xnodes(u)

# output
4-element view(OffsetArray(::StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}, -2:7), 1:4) with eltype Float64:
 0.0
 0.25
 0.5
 0.75
```

Notice that the first node is at `x=0`, the left end of the grid.
Because the `x`-direction is `Periodic`, however, the last node is `x=0.75` rather than the right end of the domain at `x=1`.
This reflects the fact that `x=0` and `x=1` are the same location.

> In a `Bounded` direction, however, the left and right end points differ and are included among the nodes.

Reviewer: ```suggestion
In a `Bounded` direction, however, the left and right end points differ and are both included among the nodes.
```

For example,

```jldoctest fields
w = Field{Center, Center, Face}(grid)

znodes(w)

# output
5-element view(OffsetArray(::StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}, -2:8), 1:5) with eltype Float64:
 0.0
 0.25
 0.5
 0.75
 1.0
```

Let's visualize the situation:

```
using CairoMakie
CairoMakie.activate!(type = "svg") # hide

fig = Figure(size=(400, 120))
ax = Axis(fig[1, 1], xlabel="x")

# Visualize the domain
lines!(ax, [0, 1], [0, 0], color=:gray)

xc = xnodes(c)
xu = xnodes(u)

scatter!(ax, xc, 0 * xc, marker=:circle, markersize=10, label="Cell centers")
scatter!(ax, xu, 0 * xu, marker=:vline, markersize=20, label="Cell interfaces")

ylims!(ax, -1, 1)
xlims!(ax, -0.1, 1.1)
hideydecorations!(ax)
hidexdecorations!(ax, ticklabels=false, label=false)
hidespines!(ax)

Legend(fig[0, 1], ax, nbanks=2, framevisible=false)

> display(fig)
> 
> save("plot_staggered_nodes.svg", fig); nothing # hide
> ```
> 
> ![](plot_staggered_nodes.svg)

Reviewer: this would do the job...

```suggestion
current_figure() # hide
```

set!(one_d_c, more_fun_stuff)

# output
7×1×1 Field{Center, Center, Center} on RectilinearGrid on CPU
├── grid: 7×1×1 RectilinearGrid{Float64, Periodic, Flat, Flat} on CPU with 3×0×0 halo
├── boundary conditions: FieldBoundaryConditions
│   └── west: Periodic, east: Periodic, south: Nothing, north: Nothing, bottom: Nothing, top: Nothing, immersed: ZeroFlux
└── data: 13×1×1 OffsetArray(::Array{Float64, 3}, -2:10, 1:1, 1:1) with eltype Float64 with indices -2:10×1:1×1:1
    └── max=19.5, min=1.5, mean=10.5
```

Now, since

```jldoctest setting
xnodes(c)

# output
4-element view(OffsetArray(::StepRangeLen{Float64, Base.TwicePrecision{Float64}, Base.TwicePrecision{Float64}, Int64}, -2:7), 1:4) with eltype Float64:
 0.5
 1.5
 2.5
 3.5
```

we find that

```jldoctest setting
c[1:4, 1, 1]

# ouptut
4-element Vector{Float64}:
 1.0
 3.0
 5.0
 7.0
```

In other words, `fun_stuff` is evaluated using the coordinates at the center of each cell for a given `Field`.
One consequence is that the result is different for a `Field` located at `x` faces,

```jldoctest setting
u = XFaceField(grid)
set!(u, fun_stuff)
u[1:4, 1, 1]

# output
4-element Vector{Float64}:
 0.0
 2.0
 4.0
 6.0

> ```

Reviewer: I felt that this section should come _before_ the bit: "For `Field`s on three-dimensional grids, the function...."

Author: Let me think about that. Also the next section is going to go into halo regions and this leads naturally into that. Maybe.

Author: I see what you're saying

Reviewer: If we keep it as is, then perhaps half a phrase to ensure the people know that now we are connecting to the example one up... E.g. "Going back to the case where we set `c` with 2x, now, since...."
